### PR TITLE
session and csrf...

### PR DESCRIPTION
**Pull request by D-360**
HttpSender script for full session and csrf token management (nashorn based). Tested over DVWA and over a real life project with active defense enabled (ESAPI)

### CODE REVIEW

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

File: httpsender/full-session-n-csrf-nashorn.js [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+		case HttpSender.BEAN_SHELL_INITIATOR: return true;
+		case HttpSender.CHECK_FOR_UPDATES_INITIATOR: return true;
+		//case HttpSender.FORCED_BROWSE_INITIATOR: return false;
+		//case HttpSender.FUZZER_INITIATOR: return false;
+		//case HttpSender.MANUAL_REQUEST_INITIATOR: return false;
+		case HttpSender.PROXY_INITIATOR: return true;
+		//case HttpSender.SPIDER_INITIATOR: return false;
+		default: return false;
+	}
+}
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:112]: Remove the lines/comments for the unnecessary entries?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+ * @param msg will be modified (not cloned).
+ */
+function doRequest(msg, helper, follow_redirects, retries, note) {
+	var requestHeader = msg.getRequestHeader();
+	var httpSender = helper.getHttpSender();
+
+	msg.setNote(note);	
+	//if(retries != null) // java.lang.IllegalStateException?
+		//httpSender.setMaxRetriesOnIOError(retries);
+
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:146]: Remove this?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+		return null;
+	}
+	return msg;
+}
+
+function sendingRequest(msg, initiator, helper) {
+
+	//print(msg.toEventData());
+
+	// skips: ACCESS_CONTROL_SCANNER_INITIATOR, AUTHENTICATION_INITIATOR, BEAN_SHELL_INITIATOR, CHECK_FOR_UPDATES_INITIATOR, PROXY_INITIATOR
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:160]: Remove this, or relocate the skips comment to the initiator check above?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+	requestHeader.setContentLength(requestBody.length());
+
+	log("{ method: sendingRequest, auth-pool: " + authPool + " }");
+	log("{ method: sendingRequest (out), path: " + msg.getRequestHeader().getPrimeHeader() + " }");
+}
+
+function responseReceived(msg, initiator, helper) {
+
+	//print(msg.toEventData());
+
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:197]: Remove?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+	if(cookies)
+		header.setCookies(cookies);
+	if(secure === true || secure === false)
+		header.setSecure(secure);
+	if(user_agent)
+		header.setDefaultUserAgent(user_agent);
+
+	//print("header: " + header);
+	//print("body: " + request.getRequestBody());
+
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:261]: Remove?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+	return request;
+}
+
+function buildNewMessage(version, secure, uri, escaped, method, data, charset, encode, content_type, cookies, user_agent) {
+	return buildFromMessage(new HttpMessage(), version, secure, uri, escaped, method, data, charset, encode, content_type, cookies, user_agent);
+}
+
+function buildPreAuthMessage() {
+	//return buildNewMessage(HTTP_VERSION, true, PRE_AUTH_GET, false, HttpRequestHeader.GET) //***
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:270]: Remove?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+	return buildNewMessage(HTTP_VERSION, null, PRE_AUTH_GET, false, HttpRequestHeader.GET)
+}
+
+function buildAuthMessage(session, csrf) {
+	var cookie = new HttpCookie(SESSION_COOKIE_NAME, session);
+	var cookies = [cookie];
+	var message = buildNewMessage(
+		HTTP_VERSION,
+		null, //true, //***
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:279]: Remove the comments?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+
+/**
+ * @param msg: final (won't be modified)
+ */
+function buildRequestCSRF(msg, session) {
+	var cookie = new HttpCookie(SESSION_COOKIE_NAME, session);
+	var cookies = [cookie];
+	var message = buildNewMessage(
+		HTTP_VERSION,
+		null, //true, //***
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:301]: Remove the comments?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+ * //print("does not remove this way! " + requestHeader.getHttpCookies().remove(cookie));
+ * //print("does not remove this way! " + requestHeader.getCookieParams().remove(cookie));
+ * 
+ */
+function addOrReplaceSessionCookie(msg, session) {
+
+	log("{ method: addOrReplaceSessionCookie, replacing sessionCookie }");
+
+	var requestHeader = msg.getRequestHeader();
+	var cookies = requestHeader.getHttpCookies(); // getCookieParams
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:411]: Remove the comment?

File: httpsender/full-session-n-csrf-nashorn.js [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+		if(current.getName() == SESSION_COOKIE_NAME) {
+			current.setValue(session);
+			notFound = false;
+		}
+	}
+
+	if(notFound)
+		cookies.add(new HttpCookie(SESSION_COOKIE_NAME, session));
+
+	requestHeader.setCookies(cookies); // req
```
Comment (kingthorin) [inline httpsender/full-session-n-csrf-nashorn.js:426]: Remove the comment?

**Review by kingthorin — COMMENTED** [review]
I have no way to re-test this. I don't see anything in the code that seems horrible, aside from the fact that it takes a rather huge amount of code. However that's neither here nor there. Most of my feedback is that some of the white space and comments can likely be removed as they're not likely relevant to end users.

**Comment by thc202** [comment]
Could you remove the commits already merged?

**Comment by thc202** [comment]
This seems to duplicate existing core functionality though.

**Comment by D-360** [comment]
Hi @thc202, hard or soft?

**Comment by D-360** [comment]
BTW, I guess you talk about the c79a9a5 commit, isn't it?

**Comment by thc202** [comment]
I was referring to 088b663, though everything can be squashed and the rebased. Note that the changelog should be updated as well.

**Comment by D-360** [comment]
Done.

**Comment by D-360** [comment]
Hello, is this PR going to be merge someday?

**Comment by D-360** [comment]
But, that was already addressed at following commits, isn't it?

**Comment by D-360** [comment]
There is no more 088b663c3ffa8ad6dd3362bc495e69121d2bfd3b at the commits' tab.

**Comment by D-360** [comment]
Ops, ok, then I'll do it later. My fault.

**Comment by kingthorin** [comment]
No worries 😉

**Comment by D-360** [comment]
Done. It's it?

**Comment by kingthorin** [comment]
Well you’ve still ended up with two commits because of merge vs rebase. I’ll try to get someone else to review then I can address whatever is left (if anything).

**Comment by D-360** [comment]
Well, I tried the rebase, but it seems I did something wrong. Sorry, and thanks.

**Comment by D-360** [comment]
Yeap, sorry, and thanks.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
The changelog needs to be updated.

**Comment by kingthorin** [comment]
Okay, I'll take care of that. It seems there's an entry, just in the wrong place.

**Comment by kingthorin** [comment]
Flattened and CHANGELOG.md tweaked.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you both!

**Review by kingthorin — APPROVED** [review]

**Pull request merged**